ledger/applications.go

```diff
@@ -235,3 +237,66 @@ func (al *logicLedger) DelGlobal(key string) error {
 func (al *logicLedger) GetDelta(txn *transactions.Transaction) (evalDelta basics.EvalDelta, err error) {
 	return al.cow.BuildEvalDelta(al.aidx, txn)
 }
+
+func (al *logicLedger) balances() (apply.Balances, error) {
+	balances, ok := al.cow.(apply.Balances)
+	if !ok {
+		return nil, fmt.Errorf("cannot get a Balances object from %v", al)
+	}
+	return balances, nil
+}
+
+func (al *logicLedger) pay(from basics.Address, to basics.Address, amount uint64, spec transactions.SpecialAddresses) error {
+	pay := transactions.PaymentTxnFields{
+		Receiver: to,
+		Amount:   basics.MicroAlgos{Raw: amount},
+	}
+	header := transactions.Header{
+		Sender: from,
+	}
+	balances, err := al.balances()
+	if err != nil {
+		return err
+	}
+	var ad transactions.ApplyData
+	err = apply.Payment(pay, header, balances, spec, &ad)
+	if err != nil {
+		return err
+	}
+	var empty transactions.ApplyData
+	if !ad.Equal(empty) {
+		return fmt.Errorf("Pay caused ad change %v", ad)
+	}
+	return nil
+}
+
+func (l *logicLedger) Perform(tx *transactions.Transaction, spec transactions.SpecialAddresses) error {
+	balances, err := l.balances()
```

Comment:
This is the heart of the matter, in the real (not mock) ledger.  There's work to be done here, particularly in how the effects are communicated, by packing them into ApplyData.